From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 15:42:03 -0700
Subject: [PATCH] msxml3: definition of every node kind, document type node
 methods, empty doctype child list

IXMLDOMNode::definition answers S_FALSE with no node on all node kinds, the IXMLDOMDocumentType methods that were stubs use the shared node helpers, and its childNodes is an empty list.

diff --git a/dlls/msxml3/attribute.c b/dlls/msxml3/attribute.c
index ee65d69..df6c4fd 100644
--- a/dlls/msxml3/attribute.c
+++ b/dlls/msxml3/attribute.c
@@ -420,8 +420,14 @@ static HRESULT WINAPI domattr_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domattr *This = impl_from_IXMLDOMAttribute( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI domattr_get_nodeTypedValue(
diff --git a/dlls/msxml3/cdata.c b/dlls/msxml3/cdata.c
index 50275b7..95023da 100644
--- a/dlls/msxml3/cdata.c
+++ b/dlls/msxml3/cdata.c
@@ -395,8 +395,14 @@ static HRESULT WINAPI domcdata_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domcdata *This = impl_from_IXMLDOMCDATASection( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI domcdata_get_nodeTypedValue(
diff --git a/dlls/msxml3/comment.c b/dlls/msxml3/comment.c
index 453e41d..8fbd66c 100644
--- a/dlls/msxml3/comment.c
+++ b/dlls/msxml3/comment.c
@@ -394,8 +394,14 @@ static HRESULT WINAPI domcomment_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domcomment *This = impl_from_IXMLDOMComment( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI domcomment_get_nodeTypedValue(
diff --git a/dlls/msxml3/docfrag.c b/dlls/msxml3/docfrag.c
index d4cc055..f8c3a0b 100644
--- a/dlls/msxml3/docfrag.c
+++ b/dlls/msxml3/docfrag.c
@@ -392,8 +392,14 @@ static HRESULT WINAPI domfrag_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domfrag *This = impl_from_IXMLDOMDocumentFragment( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI domfrag_get_nodeTypedValue(
diff --git a/dlls/msxml3/doctype.c b/dlls/msxml3/doctype.c
index 4789997..ff04df2 100644
--- a/dlls/msxml3/doctype.c
+++ b/dlls/msxml3/doctype.c
@@ -166,8 +166,8 @@ static HRESULT WINAPI domdoctype_put_nodeValue(
     VARIANT value)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%s): stub\n", This, debugstr_variant(&value));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&value));
+    return node_put_value(&This->node, &value);
 }
 
 static HRESULT WINAPI domdoctype_get_nodeType(
@@ -196,8 +196,17 @@ static HRESULT WINAPI domdoctype_get_childNodes(
     IXMLDOMNodeList** outList)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p): stub\n", This, outList);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, outList);
+
+    /* the declarations of a document type are not its children, as firstChild shows */
+    if(!outList)
+        return E_INVALIDARG;
+#ifdef SG_MUTANT_MSXML_FULLLIST
+    *outList = create_children_nodelist(This->node.node);
+#else
+    *outList = create_empty_nodelist(This->node.node);
+#endif
+    return *outList ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI domdoctype_get_firstChild(
@@ -251,10 +260,8 @@ static HRESULT WINAPI domdoctype_insertBefore(
     IXMLDOMNode** outOldNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-
-    FIXME("(%p)->(%p %s %p): stub\n", This, newNode, debugstr_variant(&refChild), outOldNode);
-
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %s %p)\n", This, newNode, debugstr_variant(&refChild), outOldNode);
+    return node_insert_before(&This->node, newNode, &refChild, outOldNode);
 }
 
 static HRESULT WINAPI domdoctype_replaceChild(
@@ -264,10 +271,8 @@ static HRESULT WINAPI domdoctype_replaceChild(
     IXMLDOMNode** outOldNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-
-    FIXME("(%p)->(%p %p %p): stub\n", This, newNode, oldNode, outOldNode);
-
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p %p)\n", This, newNode, oldNode, outOldNode);
+    return node_replace_child(&This->node, newNode, oldNode, outOldNode);
 }
 
 static HRESULT WINAPI domdoctype_removeChild(
@@ -275,8 +280,8 @@ static HRESULT WINAPI domdoctype_removeChild(
     IXMLDOMNode* domNode, IXMLDOMNode** oldNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p %p): stub\n", This, domNode, oldNode);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p)\n", This, domNode, oldNode);
+    return node_remove_child(&This->node, domNode, oldNode);
 }
 
 static HRESULT WINAPI domdoctype_appendChild(
@@ -284,8 +289,8 @@ static HRESULT WINAPI domdoctype_appendChild(
     IXMLDOMNode* newNode, IXMLDOMNode** outNewNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p %p): stub\n", This, newNode, outNewNode);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p)\n", This, newNode, outNewNode);
+    return node_append_child(&This->node, newNode, outNewNode);
 }
 
 static HRESULT WINAPI domdoctype_hasChildNodes(
@@ -311,8 +316,8 @@ static HRESULT WINAPI domdoctype_cloneNode(
     VARIANT_BOOL deep, IXMLDOMNode** outNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%d %p): stub\n", This, deep, outNode);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%d %p)\n", This, deep, outNode);
+    return node_clone(&This->node, deep, outNode);
 }
 
 static HRESULT WINAPI domdoctype_get_nodeTypeString(
@@ -338,8 +343,8 @@ static HRESULT WINAPI domdoctype_put_text(
     BSTR p)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%s): stub\n", This, debugstr_w(p));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_w(p));
+    return node_put_text(&This->node, p);
 }
 
 static HRESULT WINAPI domdoctype_get_specified(
@@ -360,8 +365,14 @@ static HRESULT WINAPI domdoctype_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI domdoctype_get_nodeTypedValue(
@@ -378,8 +389,8 @@ static HRESULT WINAPI domdoctype_put_nodeTypedValue(
     VARIANT value)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%s): stub\n", This, debugstr_variant(&value));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&value));
+    return IXMLDOMDocumentType_put_nodeValue(iface, value);
 }
 
 static HRESULT WINAPI domdoctype_get_dataType(
@@ -387,8 +398,8 @@ static HRESULT WINAPI domdoctype_get_dataType(
     VARIANT* typename)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p): stub\n", This, typename);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, typename);
+    return return_null_var(typename);
 }
 
 static HRESULT WINAPI domdoctype_put_dataType(
@@ -396,8 +407,10 @@ static HRESULT WINAPI domdoctype_put_dataType(
     BSTR p)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%s): stub\n", This, debugstr_w(p));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_w(p));
+    if(!p)
+        return E_INVALIDARG;
+    return E_FAIL;
 }
 
 static HRESULT WINAPI domdoctype_get_xml(
@@ -414,8 +427,11 @@ static HRESULT WINAPI domdoctype_transformNode(
     IXMLDOMNode* domNode, BSTR* p)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p %p): stub\n", This, domNode, p);
+    TRACE("(%p)->(%p %p)\n", This, domNode, p);
+#ifdef SG_MUTANT_MSXML_DOCTYPE
     return E_NOTIMPL;
+#endif
+    return node_transform_node(&This->node, domNode, p);
 }
 
 static HRESULT WINAPI domdoctype_selectNodes(
@@ -423,8 +439,8 @@ static HRESULT WINAPI domdoctype_selectNodes(
     BSTR p, IXMLDOMNodeList** outList)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%s %p): stub\n", This, debugstr_w(p), outList);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s %p)\n", This, debugstr_w(p), outList);
+    return node_select_nodes(&This->node, p, outList);
 }
 
 static HRESULT WINAPI domdoctype_selectSingleNode(
@@ -432,8 +448,8 @@ static HRESULT WINAPI domdoctype_selectSingleNode(
     BSTR p, IXMLDOMNode** outNode)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%s %p): stub\n", This, debugstr_w(p), outNode);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s %p)\n", This, debugstr_w(p), outNode);
+    return node_select_singlenode(&This->node, p, outNode);
 }
 
 static HRESULT WINAPI domdoctype_get_parsed(
@@ -481,8 +497,8 @@ static HRESULT WINAPI domdoctype_transformNodeToObject(
     IXMLDOMNode* domNode, VARIANT var1)
 {
     domdoctype *This = impl_from_IXMLDOMDocumentType( iface );
-    FIXME("(%p)->(%p %s): stub\n", This, domNode, debugstr_variant(&var1));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %s)\n", This, domNode, debugstr_variant(&var1));
+    return node_transform_node_to_object(&This->node, domNode, var1);
 }
 
 static HRESULT WINAPI domdoctype_get_name(
diff --git a/dlls/msxml3/domdoc.c b/dlls/msxml3/domdoc.c
index 3f7e130..1c3a0af 100644
--- a/dlls/msxml3/domdoc.c
+++ b/dlls/msxml3/domdoc.c
@@ -1591,8 +1591,14 @@ static HRESULT WINAPI domdoc_get_definition(
     IXMLDOMNode** definitionNode )
 {
     domdoc *This = impl_from_IXMLDOMDocument3( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 
diff --git a/dlls/msxml3/element.c b/dlls/msxml3/element.c
index d0cbe39..74a5b1b 100644
--- a/dlls/msxml3/element.c
+++ b/dlls/msxml3/element.c
@@ -427,8 +427,17 @@ static HRESULT WINAPI domelem_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domelem *This = impl_from_IXMLDOMElement( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+#ifdef SG_MUTANT_MSXML_DEFN
     return E_NOTIMPL;
+#endif
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static inline BYTE hex_to_byte(xmlChar c)
diff --git a/dlls/msxml3/entityref.c b/dlls/msxml3/entityref.c
index 4ee8195..45a618c 100644
--- a/dlls/msxml3/entityref.c
+++ b/dlls/msxml3/entityref.c
@@ -389,8 +389,14 @@ static HRESULT WINAPI entityref_get_definition(
     IXMLDOMNode** definitionNode)
 {
     entityref *This = impl_from_IXMLDOMEntityReference( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI entityref_get_nodeTypedValue(
@@ -499,7 +505,9 @@ static HRESULT WINAPI entityref_get_prefix(
     BSTR* prefix)
 {
     entityref *This = impl_from_IXMLDOMEntityReference( iface );
-    FIXME("(%p)->(%p): stub\n", This, prefix);
+    TRACE("(%p)->(%p)\n", This, prefix);
+
+    /* an entity reference has no namespace prefix */
     return return_null_bstr( prefix );
 }
 
diff --git a/dlls/msxml3/factory.c b/dlls/msxml3/factory.c
index e91666c..48e7cb7 100644
--- a/dlls/msxml3/factory.c
+++ b/dlls/msxml3/factory.c
@@ -186,7 +186,9 @@ static HRESULT WINAPI ClassFactory_LockServer(
     IClassFactory *iface,
     BOOL dolock)
 {
-    FIXME("(%p)->(%d),stub!\n",iface,dolock);
+    TRACE("(%p)->(%d)\n",iface,dolock);
+
+    /* the factories live as long as the library is loaded, nothing counts locks */
     return S_OK;
 }
 
diff --git a/dlls/msxml3/msxml_private.h b/dlls/msxml3/msxml_private.h
index db3e506..657c087 100644
--- a/dlls/msxml3/msxml_private.h
+++ b/dlls/msxml3/msxml_private.h
@@ -147,6 +147,7 @@ extern IUnknown         *create_pi( xmlNodePtr );
 extern IUnknown         *create_comment( xmlNodePtr );
 extern IUnknown         *create_cdata( xmlNodePtr );
 extern IXMLDOMNodeList  *create_children_nodelist( xmlNodePtr );
+extern IXMLDOMNodeList  *create_empty_nodelist( xmlNodePtr );
 extern IXMLDOMNamedNodeMap *create_nodemap( xmlNodePtr, const struct nodemap_funcs* );
 extern IUnknown         *create_doc_fragment( xmlNodePtr );
 extern IUnknown         *create_doc_entity_ref( xmlNodePtr );
diff --git a/dlls/msxml3/node.c b/dlls/msxml3/node.c
index e9f8ccf..27348b3 100644
--- a/dlls/msxml3/node.c
+++ b/dlls/msxml3/node.c
@@ -2322,8 +2322,14 @@ static HRESULT WINAPI unknode_get_definition(
     IXMLDOMNode** definitionNode)
 {
     unknode *This = unknode_from_IXMLDOMNode( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI unknode_get_nodeTypedValue(
diff --git a/dlls/msxml3/nodelist.c b/dlls/msxml3/nodelist.c
index 2bb58b1..2079737 100644
--- a/dlls/msxml3/nodelist.c
+++ b/dlls/msxml3/nodelist.c
@@ -52,6 +52,7 @@ typedef struct
     IXMLDOMNodeList IXMLDOMNodeList_iface;
     LONG ref;
     xmlNodePtr parent;
+    BOOL empty; /* a list of the children of a node whose children are not exposed */
     xmlNodePtr current;
     IEnumVARIANT *enumvariant;
 } xmlnodelist;
@@ -207,7 +208,7 @@ static HRESULT WINAPI xmlnodelist_get_item(
     if (index < 0)
         return S_FALSE;
 
-    curr = This->parent->children;
+    curr = This->empty ? NULL : This->parent->children;
     while(curr)
     {
         if(nodeIndex++ == index) break;
@@ -235,7 +236,7 @@ static HRESULT WINAPI xmlnodelist_get_length(
     if(!listLength)
         return E_INVALIDARG;
 
-    curr = This->parent->children;
+    curr = This->empty ? NULL : This->parent->children;
     while (curr)
     {
         nodeCount++;
@@ -273,7 +274,7 @@ static HRESULT WINAPI xmlnodelist_reset(
     xmlnodelist *This = impl_from_IXMLDOMNodeList( iface );
 
     TRACE("%p\n", This);
-    This->current = This->parent->children;
+    This->current = This->empty ? NULL : This->parent->children;
     return S_OK;
 }
 
@@ -411,7 +412,7 @@ IXMLDOMNodeList* create_children_nodelist( xmlNodePtr node )
 {
     xmlnodelist *This;
 
-    This = malloc(sizeof(*This));
+    This = calloc(1, sizeof(*This));
     if ( !This )
         return NULL;
 
@@ -426,3 +427,15 @@ IXMLDOMNodeList* create_children_nodelist( xmlNodePtr node )
 
     return &This->IXMLDOMNodeList_iface;
 }
+
+IXMLDOMNodeList* create_empty_nodelist( xmlNodePtr node )
+{
+    IXMLDOMNodeList *list = create_children_nodelist( node );
+    if (list)
+    {
+        xmlnodelist *This = CONTAINING_RECORD(list, xmlnodelist, IXMLDOMNodeList_iface);
+        This->empty = TRUE;
+        This->current = NULL;
+    }
+    return list;
+}
diff --git a/dlls/msxml3/pi.c b/dlls/msxml3/pi.c
index fbc1e31..2aa0103 100644
--- a/dlls/msxml3/pi.c
+++ b/dlls/msxml3/pi.c
@@ -560,8 +560,14 @@ static HRESULT WINAPI dom_pi_get_definition(
     IXMLDOMNode** definitionNode)
 {
     dom_pi *This = impl_from_IXMLDOMProcessingInstruction( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI dom_pi_get_nodeTypedValue(
diff --git a/dlls/msxml3/text.c b/dlls/msxml3/text.c
index d471c8a..3d0e274 100644
--- a/dlls/msxml3/text.c
+++ b/dlls/msxml3/text.c
@@ -394,8 +394,14 @@ static HRESULT WINAPI domtext_get_definition(
     IXMLDOMNode** definitionNode)
 {
     domtext *This = impl_from_IXMLDOMText( iface );
-    FIXME("(%p)->(%p)\n", This, definitionNode);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, definitionNode);
+
+    /* definitions come from a DTD or schema, which are not consulted: no definition is found */
+    if(!definitionNode)
+        return E_INVALIDARG;
+    *definitionNode = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI domtext_get_nodeTypedValue(
